### ISSUE TITLE
Cells with conditional formulas with an empty result return 0 instead of an empty string

### ISSUE THREAD

**Issue by mspacek** [python-excel/xlrd#207]
Hello! I have an .xls sheet with a formula that's been extended part way down a column. The formula, say in cell C9, has a conditional, something of the form:
```
IF(A9="foo", 2*B9, "")
```
Most of the time, column A is not equal to "foo", and therefore column C remains empty. However, xlrd returns such cells as having an integer value of 0. This is unexpected behaviour, and causes downstream problems in Pandas. In this particular case, all of column C is formatted as "General Number". Note that the rows in column C to which the formula has not been extended are correctly returned as empty strings.

Is there some way to make xlrd only consider the actual value of a cell, as displayed in Excel or Libreoffice Calc, instead of considering whether or not it has a formula? Apparently, openpyxl has a kwarg called `data_only=True` that does exactly this:

http://stackoverflow.com/a/28100021
http://openpyxl.readthedocs.io/en/default/usage.html#read-an-existing-workbook

If xlrd can't currently do this, would it be hard to implement? I could try and give it a shot, if pointed in the right direction.

**Comment by sjmachin** [comment]
Re Most of the time, column C is not equal to "foo" -- don't you mean column A?

Re therefore [col C] remains empty -- it's not empty, the cell type is "text" and the value is the zero-length string.

xlrd takes no notice of the formula. What it gives you is what the cell contained when the file was saved. If the file was produced by software that has no formula evaluater (e.g. xlwt), you will get a default value, typically zero. Try opening the file in Excel and saving it again; this should recalculate the formula values.

openpyxl works on xlsx files, not xls. Totally different formula mechanism.

**Comment by mspacek** [comment]
Sorry for the delay. I didn't get an email notification for some reason.

> Re Most of the time, column C is not equal to "foo" -- don't you mean column A?

Sorry, yes, that's what I meant. I've updated the original comment.

I assumed that "empty" and zero-length string were the same thing, but apparently they are not:

http://stackoverflow.com/questions/18768020/if-statement-how-to-leave-cell-blank-if-condition-is-false-does-not-work

> Unfortunately, there is no formula way to result in a truly blank cell, "" is the best formulas can offer.

Here is the output from xlrd (slightly reformatted for readability) of a fairly minimal example of the .xls sheet described in the original comment. It has 6 rows of data. The first 4 rows of column C have the equation `=IF(A2="foo", 2*B2, "")`, with the row number incrementing appropriately. The middle two rows of column C evaluate to the empty string, but fail `ISBLANK()`. These two rows are displayed as blank in both Excel and LO Calc, but xlrd returns them as 0.0. The last 2 rows of column C have no equation in them, pass `ISBLANK()`, are displayed as blank in Excel and LO Calc, and xlrd returns them as an empty Python string.
```
['A',    'B', 'C', 'C is emptystr', 'C is blank']
['foo',  1.0, 2.0,      0.0,             0      ]
['foo',  2.0, 4.0,      0.0,             0.0    ]
['blah', 3.0, 0.0,      1.0,             0.0    ]
['blah', 4.0, 0.0,      1.0,             0.0    ]
['blah', 5.0,  '',      1.0,             1.0    ]
['blah', 6.0,  '',      1.0,             1.0    ]
```
So in summary, when xlrd encounters a cell with an empty string `""`, it returns 0. Wouldn't it make more sense for it to do what it does when it encounters a truly empty cell, i.e. return an empty Python string?

I've attached a .zip file with the code for the above, a screenshot of the .xls sheet in LO Calc, and the .xls file itself: [test_xlrd.zip](https://github.com/python-excel/xlrd/files/855940/test_xlrd.zip)

**Comment by mspacek** [comment]
Well, I guess I've found a bug in LO Calc (Version: 5.3.1.2 Build ID: 1:5.3.1-0ubuntu1~xenial0). After adding a bunch of debugging messages to `sheet.read()`, I found that the above .xls file saved in LO Calc doesn't set the formula result type properly to string, i.e., `result_str[6:8] == b"\xFF\xFF"` is `False` around line 927. In fact, the whole `result_str` for the fields with what should be formula results with empty strings are nothing but null bytes, which are then interpreted as the number 0. When I resave the above .xls file in Excel (2016, Mac), I get this instead:
```
['A',    'B', 'C', 'C is emptystr', 'C is blank']
['foo',  1.0, 2.0,        0,             0      ]
['foo',  2.0, 4.0,        0,             0      ]
['blah', 3.0,  '',        1,             0      ]
['blah', 4.0,  '',        1,             0      ]
['blah', 5.0,  '',        1,             1      ]
['blah', 6.0,  '',        1,             1      ]
```
which is exactly what I'd expect! I also just discovered that if I fiddle with the original file a bit in LO Calc, replacing the empty string in the formula in column C with `NA()`, extending it across the 4 rows, undoing the changes, and then saving, I also get what I expect, nearly the identical result as from Excel:
```
['A',    'B', 'C', 'C is emptystr', 'C is blank']
['foo',  1.0, 2.0,       0.0,            0      ]
['foo',  2.0, 4.0,       0.0,            0      ]
['blah', 3.0,  '',       1.0,            0      ]
['blah', 4.0,  '',       1.0,            0      ]
['blah', 5.0,  '',       1.0,            1      ]
['blah', 6.0,  '',       1.0,            1      ]
```
What a mess. I guess there's something finicky going on in LO Calc, and I'm not quite sure what. Sorry for the noise. For future reference, here are the three versions of the same .xls file. The one named `empty_cells_test_lo_error.xls` is the one causing trouble. The other two both work fine: [empty_cells_test.zip](https://github.com/python-excel/xlrd/files/856618/empty_cells_test.zip)

**Comment by sjmachin** [comment]
I have opened your xls file in Excel 2003 and saved it without change as xls. The result when read by xlrd is that cells C4 and C5 (originally number type and value = 0.0) are now as expected (type = text and value = zero-length string). It is not an xlrd problem -- the files are different.
